saga/sagaRecovery.go

```diff
@@ -0,0 +1,101 @@
+package saga
+
+import (
+	"fmt"
+)
+
+type SagaRecoveryType int
+
+/*
+ * Saga Recovery Types define how to interpret SagaState in RecoveryMode.
+ *
+ * ForwardRecovery: all tasks in the saga must be executed at least once.
+ *                  tasks MUST BE idempotent
+ *
+ * RollbackRecovery: if Saga is Aborted or in unsafe state, compensating
+ *                   tasks for all started tasks need to be executed.
+ *                   compensating tasks MUST BE idempotent.
+ */
+const (
+	RollbackRecovery SagaRecoveryType = iota
+	ForwardRecovery
+)
+
+/*
+ * Recovers SagaState from SagaLog messages
+ */
+func recoverState(sagaId string, saga Saga, recoveryType SagaRecoveryType) (*SagaState, error) {
+
+	// Get Logged Messages For this Saga from the Log.
+	msgs, err := saga.log.GetMessages(sagaId)
+	if err != nil {
+		return nil, err
+	}
+
+	if msgs == nil || len(msgs) == 0 {
+		return nil, nil
+	}
+
+	// Reconstruct Saga State from Logged Messages
+	startMsg := msgs[0]
+	if startMsg.msgType != StartSaga {
+		return nil, fmt.Errorf("InvalidMessages: first message must be StartSaga")
+	}
+
+	state, err := makeSagaState(sagaId, startMsg.data)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, msg := range msgs {
+		state, err = updateSagaState(state, msg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Check if we can safely proceed forward based on recovery method
+	// RollbackRecovery must check if in a SafeState,
+	// ForwardRecovery can always make progress
+	switch recoveryType {
+
+	case RollbackRecovery:
```

Comment:
I don't really understand why we're modifying SagaState here, that's more than just recovery. Also, how does the caller know if it should call compensating tasks due to an abort here (if that's the intent), and what is the expected action for a logged abort vs one we've applied right here?